src/s3/async_put_object.rs

```diff
@@ -0,0 +1,232 @@
+use aws_sdk_s3::error::PutObjectError;
+use aws_sdk_s3::model::ObjectCannedAcl;
+use aws_sdk_s3::output::PutObjectOutput;
+use aws_sdk_s3::types::SdkError;
+use futures::io::{Error, ErrorKind};
+use futures::task::{Context, Poll};
+use futures::{AsyncWrite, Future};
+use std::mem;
+use std::pin::Pin;
+
+use crate::s3::Client;
+
+// Tracks the state of the AsyncWrite lifecycle for an AsyncPutObject.
+enum PutObjectState<'a> {
+    // Open for writing. Can call .write(), .flush(), or .close().
+    Writing,
+    // In the process of writing the data to s3. We store the future which is performing
+    // the pub_object operation.
+    Closing(Pin<Box<dyn Future<Output = Result<PutObjectOutput, SdkError<PutObjectError>>> + 'a>>),
+    // We have completed writing to s3.
+    Closed,
+}
+
+/// Implements the `AsyncWrite` trait and writes data to S3 using the `put_object` API.
+///
+/// # Example
+///
+/// ```no_run
+/// use aws_config;
+/// use cobalt_aws::s3::{get_client, AsyncPutObject};
+/// use futures::AsyncWriteExt;
+///
+/// # tokio_test::block_on(async {
+/// let shared_config = aws_config::load_from_env().await;
+/// let client = get_client(&shared_config).unwrap();
+///
+/// let mut writer = AsyncPutObject::new(&client, "my-bucket", "my-key");
+/// writer.write_all(b"File contents").await.unwrap();
+///
+/// // The contents are pushed to S3 when the `.close()` method is called.
+/// writer.close().await.unwrap();
+/// # })
+/// ```
+pub struct AsyncPutObject<'a> {
+    client: &'a Client,
+    key: String,
+    bucket: String,
+    buf: Vec<u8>,
+    state: PutObjectState<'a>,
+}
+
+impl<'a> AsyncPutObject<'a> {
+    /// Create a new `AsyncPutObject` which will write data to the given `bucket` and `key`.
+    pub fn new(client: &'a Client, bucket: &str, key: &str) -> Self {
+        AsyncPutObject {
+            client,
+            key: key.into(),
+            bucket: bucket.into(),
+            buf: vec![],
+            state: PutObjectState::Writing,
+        }
+    }
+}
+
+impl<'a> AsyncWrite for AsyncPutObject<'a> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, Error>> {
+        match self.state {
+            PutObjectState::Writing => {
+                self.buf.extend(buf);
+                Poll::Ready(Ok(buf.len()))
+            }
+            _ => Poll::Ready(Err(Error::new(
+                ErrorKind::Other,
+                "Attempted to .write() writer after .close().",
+            ))),
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        match self.state {
+            PutObjectState::Writing => Poll::Ready(Ok(())),
```

Comment:
I wonder if we should error out here rather than returning `Ok()`. Flush is not going to do what the caller expects, and maybe it's better to refuse to pretend we will.

Reply:
Yeah, I wonder what the impact of that might be? My concern is that might be other libraries out there which choose to `.flush()` occasionally, but not with any hard expectations on the outcome of that operation. From our POV that's harmless enough, and it would be a PITA if we couldn't use such a tool with our object. On the other hand, perhaps better to make a noise and wait for that noise to get back to our ears to make a decision 🤔 I think for now let's ship what we have, but keep an eye on this as we go forward.